### ISSUE TITLE
Weighted stat aggregations in arrow-compute

### ISSUE THREAD

**Issue by RNKuhns** [apache/arrow#15103]
### Describe the enhancement requested

Arrow’s compute functions currently include several aggregate statistics (mean, sum, variance, etc).

It would be great to offer weighted versions of several of these (mean, sum, count, variance and standard deviation in short run, quantile in longer-run) and expose them to pyarrow, R, etc. This would allow this functionality to be pushed down to Arrow and Arrow datasets. For example, an R user would have to either collect the data first and apply something like weighted.mean from R stats or code up custom logic using functionality available in Arrow to arrive at a similar result. Eother is not ideal for a relatively routine aggregation (weighted statistics). It would also make it much easier to calculate weighted statistics when working with distributed Arrow datasets in Python.

Note that this is not covered by UDF API directly which doesn’t support aggregate functions.

The added functionality could be through the creation of functions with a signature like: weighted_mean(x: arrow array, weights: arrow array) -> scalar

### Component(s)

C++, Python, R

**Comment by westonpace** [comment]
There is a proposal to support aggregate UDFs but I don't know that it is high priority for anyone.  I agree this sounds like a nice feature.  Would you be interested in creating a PR?

**Comment by RNKuhns** [comment]
@westonpace I'm currently doing some work on another project. But I am interested in working to get this added if it fits my availability time table (about 1-2 months).

If that works, I could start by taking a look at the proposal and getting up to speed ahead of my time opening up.

I am also curious how the development on Apache Arrow and [Apache Arrow Datafusion](https://arrow.apache.org/datafusion/) (or [python bindings](https://github.com/apache/arrow-datafusion-python) work together.

I believe datafusion is built on the Arrow format and has some support for user defined aggregations. Do the projects work closely to re-use/align overlapping functionality?

**Comment by westonpace** [comment]
> I am also curious how the development on Apache Arrow and [Apache Arrow Datafusion](https://arrow.apache.org/datafusion/) (or [python bindings](https://github.com/apache/arrow-datafusion-python) work together.

Pyarrow, datasets, and the streaming execution engine (Acero) are all based on arrow-C++.  Datafusion is based on arrow-rs (rust).  So they are two separate stacks.

That being said, they should be very interoperable.  Both can import and export to the C data interface for intra-process communication.  Both can read and write the arrow IPC format for IPC over something like flight for inter-process communication.

> Do the projects work closely to re-use/align overlapping functionality?

Unfortunately, a universal UDF harness has not yet been created.  In general, datafusion, based on rust, is not likely to trust a C/C++ UDFs for safety reasons (at least, that is my understanding, I could be wrong).  Going the other way (using datafusion UDFs in pyarrow) should be possible but no one has done it yet.  Some very interesting ideas have been proposed for other alternatives (there was a discussion a while back on the mailing list).  Perhaps the most interesting to me is the idea of using WASM for UDFs.  LLVM has also been proposed though it doesn't fix the safety concerns.

**Comment by lorentzenchr** [comment]
Basic weighted statistics would be super useful. If C++ arrow compute functions could provide

- weighted mean / average
- weighted variance

The use case I have in mind is a grouped-by-aggregate. While the weighted mean is easy to do yourself - but still tedious - the weighted variance is hard to do in one `table.grouped_by(..).aggregate(..)` sequence.

**Comment by westonpace** [comment]
> Basic weighted statistics would be super useful. If C++ arrow compute functions could provide

Yes, I may have let myself get sidetracked with UDF discussions.  I think it would be great for someone to contribute these directly as well.  Support was recently added for binary aggregate kernels so it is only a matter of creating a new kernel.

**Comment by lorentzenchr** [comment]
Could you point to one binary aggregating compute function? That would help to consider contributing:smirk:

**Comment by felipecrv** [comment]
@lorentzenchr I have a drafty one that I need to push (that would be the first!). I will do it now.

**Comment by thisisnic** [comment]
This issue has been marked as stale because it has had no activity in the past 2 years. Please remove the stale label or comment below, or this issue will be closed in 14 days.
